Add config to capture stacktrace when a span duration exceeds threshold

tylerbenson opened the pull request:
(But only when span is not errored or finished on a different thread.)

Use the following config:
```
-Ddd.trace.span.duration-above-average.stacktrace.millis=1000
```
(One second is the default, `0` disables.)

dougqh commented on dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java:
```diff
@@ -86,12 +89,112 @@ private void finishAndAddToTrace(final long durationNano) {
   }
 
+  private void addStacktraceIfThresholdExceeded() {
+    final long spanDurationStacktraceNanos = Config.get().getSpanDurationStacktraceNanos();
```
Long term, I'd like to get away from the use of Singletons like Config -- but for now, I think this is fine.

dougqh commented on dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java:
```diff
@@ -86,12 +89,112 @@ private void finishAndAddToTrace(final long durationNano) {
+        && durationNano.get() > spanDurationStacktraceNanos
+        // If this span was finished async, then the stacktrace will be less meaningful.
+        && context.threadId == Thread.currentThread().getId()) {
+      final Exception stacktrace = new Exception();
```
Why not call Thread.getStackTraceElement[] instead?
Ideally, we also use the frame walking API on newer versions of Java, but that can wait until later.

dougqh commented on dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java:
```diff
@@ -86,12 +89,112 @@ private void finishAndAddToTrace(final long durationNano) {
 
+  private void addStacktraceIfThresholdExceeded() {
+    final long spanDurationStacktraceNanos = Config.get().getSpanDurationStacktraceNanos();
+    if (!isError()
```
I'd like to see a health metric for anomalous situations as well.

dougqh commented on dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java:
```diff
@@ -86,12 +89,112 @@ private void finishAndAddToTrace(final long durationNano) {
     // ensure a min duration of 1
     if (this.durationNano.compareAndSet(0, Math.max(1, durationNano))) {
       log.debug("Finished: {}", this);
+      addStacktraceIfThresholdExceeded();
```
This feels like an optional behavior that might be better handled through a Listener mechanism.
However, I'm fine with deferring that to a later PR.

dougqh commented on dd-trace-ot/src/main/java/datadog/opentracing/DDSpan.java:
```diff
@@ -86,12 +89,112 @@ private void finishAndAddToTrace(final long durationNano) {
+
+    @Override
+    public void write(final char[] cbuf, final int off, final int len) throws IOException {
+      if ((off < 0)
```
I think requesting and traversing the StackTraceElement[] would be easier to follow that filtering Writer.

jkubrynski replied:
It would be useful to make the filter configurable. There is often a lot of framework code that can be ignored. I'd even consider including only frames from my code. It worked great in `mjprof`

tylerbenson replied:
For simplicity, I'm not going to make it configurable in this PR... we can evaluate that later.

jkubrynski commented on dd-trace-api/src/main/java/datadog/trace/api/Config.java:
```diff
@@ -129,6 +132,8 @@
   private static final String DEFAULT_SPLIT_BY_TAGS = "";
   private static final int DEFAULT_PARTIAL_FLUSH_MIN_SPANS = 1000;
+  private static final int DEFAULT_SPAN_DURATION_STACKTRACE_MILLIS =
+      (int) TimeUnit.SECONDS.toMillis(1);
```
Depending on a use-case of course, but for regular web applications `1 second` is a strict setting, and many users can be flooded.

tylerbenson replied:
I was trying to find a balance between a config default that is useful, but not burdensome...  My thought is that a stacktrace is expensive, but if a span is more than a second, then the cost of a stacktrace is likely not noticeable compared to the rest of the trace.  Do you have any suggestions for a different config default?

tylerbenson commented:
@dougqh @jkubrynski, I've updated the PR with your suggestions.  Good idea about the `StackTraceElement`.  That was much easier and should have been my first solution.  I've added a todo for the future to consider using the `StackWalking` api.
